test/memspaces/memspace_host_all.cpp

```diff
@@ -0,0 +1,91 @@
+#include "memspace_helpers.hpp"
+#include "memspace_internal.h"
+#include "test_helpers.h"
+
+#include <numa.h>
+#include <numaif.h>
+#include <umf/memspace.h>
+
+using umf_test::test;
+
+TEST_F(numaNodesTest, memspaceGet) {
+    umf_memspace_handle_t hMemspace = umfMemspaceHostAllGet();
+    UT_ASSERTne(hMemspace, nullptr);
+
+    // Confirm that the HOST ALL memspace is composed of all available NUMA nodes.
+    UT_ASSERTeq(hMemspace->size, nodeIds.size());
+    for (size_t i = 0; i < hMemspace->size; i++) {
+        // NUMA memory target internally casts the config directly into priv.
+        struct umf_numa_memory_target_config_t *numaTargetCfg =
+            (struct umf_numa_memory_target_config_t *)hMemspace->nodes[i]->priv;
+        UT_ASSERT(std::find(nodeIds.begin(), nodeIds.end(),
+                            numaTargetCfg->id) != nodeIds.end());
+    }
+}
+
+TEST_F(memspaceHostAllTest, providerFromHostAllMemspace) {
+    umf_memory_provider_handle_t hProvider = nullptr;
+    enum umf_result_t ret =
+        umfMemoryProviderCreateFromMemspace(hMemspace, nullptr, &hProvider);
+    UT_ASSERTeq(ret, UMF_RESULT_SUCCESS);
+    UT_ASSERTne(hProvider, nullptr);
+
+    umfMemoryProviderDestroy(hProvider);
+}
+
+TEST_F(memspaceHostAllProviderTest, allocFree) {
+    void *ptr = nullptr;
+    size_t size = SIZE_4K;
+    size_t alignment = 0;
+
+    enum umf_result_t ret =
+        umfMemoryProviderAlloc(hProvider, size, alignment, &ptr);
+    UT_ASSERTeq(ret, UMF_RESULT_SUCCESS);
+    UT_ASSERTne(ptr, nullptr);
+
+    memset(ptr, 0xFF, size);
+
+    ret = umfMemoryProviderFree(hProvider, ptr, size);
+    UT_ASSERTeq(ret, UMF_RESULT_SUCCESS);
+}
+
+TEST_F(memspaceHostAllProviderTest, memoryPolicy) {
+    void *ptr = nullptr;
+    size_t size = SIZE_4K;
+    size_t alignment = 0;
+
+    enum umf_result_t umf_ret =
+        umfMemoryProviderAlloc(hProvider, size, alignment, &ptr);
+    UT_ASSERTeq(umf_ret, UMF_RESULT_SUCCESS);
+    UT_ASSERTne(ptr, nullptr);
+
+    auto it = std::max_element(nodeIds.begin(), nodeIds.end());
+    UT_ASSERT(it != nodeIds.end());
+
+    unsigned long maxNodeId = *it;
+
+    unsigned bitsPerUlong = sizeof(unsigned long) * 8;
+    unsigned nrUlongs = (maxNodeId + bitsPerUlong) / bitsPerUlong;
+    std::vector<unsigned long> memNodeMasks(nrUlongs, 0);
+
+    int memMode = -1;
+    int ret = get_mempolicy(&memMode, memNodeMasks.data(),
+                            nrUlongs * bitsPerUlong, ptr, MPOL_F_ADDR);
+    UT_ASSERTeq(ret, 0);
+    // Obtained memory should be bound to all available NUMA nodes.
+    UT_ASSERTeq(memMode, MPOL_BIND);
+    for (auto &nodeId : nodeIds) {
+        size_t memNodeMaskIdx = ((nodeId + bitsPerUlong) / bitsPerUlong) - 1;
+        auto &memNodeMask = memNodeMasks.at(memNodeMaskIdx);
+
+        UT_ASSERT(memNodeMask && (1UL << (nodeId % bitsPerUlong)));
+    }
+
+    umf_ret = umfMemoryProviderFree(hProvider, ptr, size);
+    UT_ASSERTeq(umf_ret, UMF_RESULT_SUCCESS);
+}
```

Comment:
what I miss here is the test that allocs memory until we got OOM and checks if all nodes are used

Reply:
Do we really need OOM test for such case? From my understanding, the provider created from memspace should always return memory that's bound to every node present on such memspace, and the `memoryPolicy` test case tests that. Would you like me to change this test to create multiple allocations?

I could create a test where memory is obtained until `UMF_OS_RESULT_ERROR_ALLOC_FAILED` OS provider native error (there's no clear OOM error) if it's necessary.

Reply:
yes, "should" :) yes, please add this case - simple loop + check if we allocate at least one allocation on each node would be enough (checking the total size of allcaction vs total size of NUMA nodes could be tricky because of swap file)